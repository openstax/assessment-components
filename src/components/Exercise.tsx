import styled from 'styled-components';
import { Content } from './Content';
import { TaskStepCard } from './Card';
import { AvailablePoints, ExerciseAnswerState, ExerciseData, ID, Step } from 'src/types';
import { ExerciseQuestion } from './ExerciseQuestion';

const StyledTaskStepCard = styled(TaskStepCard)`
  font-size: 1.8rem;
  line-height: 3rem;
`;

const Preamble = ({ exercise }: { exercise: ExerciseData }) => {
  return (
    <>
      {exercise.context &&
        <Content className="step-card-body exercise-context"
          block html={exercise.context} />}

      {exercise.stimulus_html &&
        <Content className="step-card-body exercise-stimulus"
          block html={exercise.stimulus_html} />}
    </>
  );
};

export interface ExerciseProps {
  exercise: ExerciseData;
  step?: Step;
  numberOfQuestions: number;
  questionNumber: number;
  canAnswer: boolean;
  needsSaved: boolean;
  answer_id_order?: ID[];
  hasMultipleAttempts: boolean;
  onAnswerChange: () => void;
  onAnswerSave: () => void;
  onNextStep: () => void;
  canUpdateCurrentStep: boolean;
  attempt_number: number;
  apiIsPending: boolean;
  available_points: AvailablePoints;
  displaySolution: boolean;
  show_all_feedback?: boolean;
  exerciseAnswers: {[key: string]: ExerciseAnswerState};
}

export const Exercise = ({
  displaySolution, numberOfQuestions, questionNumber, step, exercise, canAnswer, needsSaved, exerciseAnswers, ...props
}: ExerciseProps) => (
  <StyledTaskStepCard
    step={step}
    questionNumber={questionNumber}
    numberOfQuestions={numberOfQuestions}
  >
    <Preamble exercise={exercise} />

    {exercise.questions.map((q) =>
      <ExerciseQuestion
        {...props}
        {...step}
        {...exerciseAnswers[q.id]}
        exercise_uid={exercise.uid}
        key={q.id}
        question={q}
        questionNumber={questionNumber}
        choicesEnabled={canAnswer}
        canAnswer={canAnswer}
        needsSaved={needsSaved}
        canUpdateCurrentStep={canAnswer}
        displaySolution={!!displaySolution}
        answerId={step?.answer_id}
      />
    )}
  </StyledTaskStepCard>
);
