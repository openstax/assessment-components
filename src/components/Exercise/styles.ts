import { css } from "styled-components";

export const exerciseStyles = css`

  &.included-card {
    background-color: #daf3f8;
    .step-card-footer-inner, .step-card-body, .step-card-header {
      background-color: #daf3f8 !important;
    }

    .answer-letter-wrapper::before:not(.answer-correct} {
      background-color: #daf3f8 !important;
    }
  }

  &.preview-card {
    --spacing: 0.8rem;

    .step-card-header,
    .step-card-body {
      background-color: #FFFFFF;
      padding: var(--spacing);
      font-size: 1.6rem;
      line-height: 2rem;
    }

    .step-card-footer {
      padding: 0 var(--spacing) var(--spacing);
    }

    .step-card-footer-inner {
      padding-top: var(--spacing);
    }

    .answers-table {
      margin: 0;
    }

    .exercise-context {
      .os-title {
        margin: 0 0 var(--spacing);
      }
      .os-subtitle {
        margin: var(--spacing) 0;
      }
      figure {
        float: left;
        margin: 0;
      }
      img {
        max-width: 16rem;
      }
      p:last-child {
        margin-bottom: 0;
      }
    }
    .openstax-question {
      .openstax-answer {
        padding: 0;
        border: none;

        .answer-label {
          padding-top: var(--spacing);
        }

        .answer-letter-wrapper {
          .answer-answer {
            margin-left: var(--spacing);
          }
        }

        .answer-letter {
          min-width: 2.8rem;
          min-height: 2.8rem;
          width: 2.8rem;
          height: 2.8rem;
          display: flex;
        }
      }

      .question-feedback-content {
        font-size: 1.6rem;
      }

      .points,
      .question-feedback {
        max-width: unset;
      }
    }

    .step-card-footer, .detailed-solution {
      display: none;
    }

    .question-stem,
    .question-feedback-content,
    .question-info,
    .exercise-context {
      line-height: 2rem;
    }

    .question-info {
      font-weight: bold;
      font-size: 1.6rem;
    }

    .question-id {
      font-weight: 400;
      font-size: 1.6rem;
    }

    .question-stem {
      color: #424242;
      font-weight: bold;
      font-size: 1.6rem;
    }
  }
`;
