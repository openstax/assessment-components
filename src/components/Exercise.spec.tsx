import { Exercise, ExerciseProps } from './Exercise';
import renderer from 'react-test-renderer';

describe('Exercise', () => {
  let props: ExerciseProps;

  beforeEach(() => {
    props = {
      exercise: {
        uid: '1@1',
        uuid: 'e4e27897-4abc-40d3-8565-5def31795edc',
        group_uuid: '20e82bf6-232e-40c8-ba68-2d22c6498f69',
        number: 1,
        version: 1,
        published_at: '2022-09-06T20:32:21.981Z',
        context: 'Context',
        stimulus_html: '<b>Stimulus HTML</b>',
        tags: [],
        authors: [{ user_id: 1, name: 'OpenStax' }],
        copyright_holders: [{ user_id: 1, name: 'OpenStax' }],
        derived_from: [],
        is_vocab: false,
        solutions_are_public: false,
        versions: [1],
        questions: [{
          id: '1234@5',
          collaborator_solutions: [],
          formats: ['true-false'],
          stimulus_html: '',
          stem_html: '',
          is_answer_order_important: false,
          answers: [{
            id: '1',
            correctness: undefined,
            content_html: 'True',
          }, {
            id: '2',
            correctness: undefined,
            content_html: 'False',
          }],
        }],
      },
      displaySolution: false,
      exerciseAnswers: {},
      questionNumber: 1,
      hasMultipleAttempts: false,
      onAnswerChange: () => null,
      onAnswerSave: () => null,
      onNextStep: () => null,
      available_points: '1.0' as const,
      canAnswer: false,
      needsSaved: false,
      canUpdateCurrentStep: false,
      attempt_number: 0,
      apiIsPending: false,
      step: {
        type: 'exercise' as const,
        task: {
          is_deleted: false,
          type: 'homework' as const
        },
        uid: '1234@4',
        id: 1,
        available_points: '1.0',
        preview: 'Preview content',
        is_completed: false,
        answer_id_order: ['1', '2'],
        answer_id: '1',
        free_response: '',
        last_completed_at: new Date(),
        feedback_html: '',
        correct_answer_id: '',
        correct_answer_feedback_html: '',
        external_url: '',
        can_be_updated: false,
        is_feedback_available: true,
        exercise_id: '1',
        attempts_remaining: 0,
        attempt_number: 1,
        incorrectAnswerId: 0
      },
      numberOfQuestions: 1
    }
  });

  it('matches snapshot', () => {
    const tree = renderer.create(
      <Exercise {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

});
