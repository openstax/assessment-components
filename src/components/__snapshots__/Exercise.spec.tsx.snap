// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercise using step data matches snapshot 1`] = `
<div
  className="sc-gsnTZi kTXtGE"
  data-task-step-id={1}
>
  <div
    className="sc-bczRLJ hctMxl exercise-step sc-iBkjds Alzlt"
  >
    <div
      className="sc-dkzDqf ifUCJL"
    >
      <div>
        <div
          className="question-info"
        >
          <span>
            Question 1
          </span>
          <span
            className="num-questions"
          >
             / 
            1
          </span>
          <span
            className="separator"
          >
            |
          </span>
          <span
            className="question-id"
          >
            ID: 
            1234@4
          </span>
        </div>
      </div>
      <div>
        <div
          className="points"
        >
          1.0
           Points
        </div>
      </div>
    </div>
    <div
      className="sc-hKMtZM grRsJk"
    >
      <div
        className="step-card-body exercise-context"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Context",
          }
        }
      />
      <div
        className="step-card-body exercise-stimulus"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<b>Stimulus HTML</b>",
          }
        }
      />
      <div
        data-test-id="student-exercise-question"
      >
        <div
          className="sc-jSMfEi kWGKid openstax-question step-card-body"
          data-question-number={1}
          data-test-id="question"
        >
          <div
            className="answers-table"
          >
            <div
              className="openstax-answer"
            >
              <section
                className="answers-answer disabled answer-checked"
                role="region"
              >
                <input
                  checked={true}
                  className="answer-input-box"
                  disabled={true}
                  id="1234@5-option-0"
                  name="1234@5-options"
                  onChange={[Function]}
                  type="radio"
                />
                <label
                  className="answer-label"
                  htmlFor="1234@5-option-0"
                >
                  <span
                    className="answer-letter-wrapper"
                  >
                    <button
                      aria-label="Selected Choice a:"
                      className="answer-letter"
                      data-test-id="answer-choice-a"
                      disabled={true}
                      onClick={[Function]}
                    >
                      a
                    </button>
                  </span>
                  <div
                    className="answer-answer"
                  >
                    <span
                      className="answer-content"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "True",
                        }
                      }
                    />
                  </div>
                </label>
              </section>
            </div>
            <div
              className="openstax-answer"
            >
              <section
                className="answers-answer disabled"
                role="region"
              >
                <input
                  checked={false}
                  className="answer-input-box"
                  disabled={true}
                  id="1234@5-option-1"
                  name="1234@5-options"
                  onChange={[Function]}
                  type="radio"
                />
                <label
                  className="answer-label"
                  htmlFor="1234@5-option-1"
                >
                  <span
                    className="answer-letter-wrapper"
                  >
                    <button
                      aria-label="Choice b:"
                      className="answer-letter"
                      data-test-id="answer-choice-b"
                      disabled={true}
                      onClick={[Function]}
                    >
                      b
                    </button>
                  </span>
                  <div
                    className="answer-answer"
                  >
                    <span
                      className="answer-content"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "False",
                        }
                      }
                    />
                  </div>
                </label>
              </section>
            </div>
          </div>
        </div>
        <div
          className="sc-gKXOVf kGsBya step-card-footer"
        >
          <div
            className="points"
          >
            <strong>
              Points: 
              1.0
            </strong>
            <span
              className="attempts-left"
            />
          </div>
          <div
            className="controls"
          >
            <button
              className="sc-eCYdqJ dqiaKp"
              data-test-id="continue-btn"
              onClick={[Function]}
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Exercise with question state data matches snapshot 1`] = `
<div
  className="sc-gsnTZi kTXtGE"
  data-task-step-id={1}
>
  <div
    className="sc-bczRLJ hctMxl exercise-step sc-iBkjds Alzlt"
  >
    <div
      className="sc-dkzDqf ifUCJL"
    >
      <div>
        <div
          className="question-info"
        >
          <span>
            Question 1
          </span>
          <span
            className="num-questions"
          >
             / 
            1
          </span>
          <span
            className="separator"
          >
            |
          </span>
          <span
            className="question-id"
          >
            ID: 
            1234@5
          </span>
        </div>
      </div>
      <div>
        <div
          className="points"
        >
          1.0
           Points
        </div>
      </div>
    </div>
    <div
      className="sc-hKMtZM grRsJk"
    >
      <div
        className="step-card-body exercise-context"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Context",
          }
        }
      />
      <div
        className="step-card-body exercise-stimulus"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<b>Stimulus HTML</b>",
          }
        }
      />
      <div
        data-test-id="student-exercise-question"
      >
        <div
          className="sc-jSMfEi kWGKid openstax-question step-card-body"
          data-question-number={1}
          data-test-id="question"
        >
          <div
            className="answers-table"
          >
            <div
              className="openstax-answer"
            >
              <section
                className="answers-answer disabled answer-checked"
                role="region"
              >
                <input
                  checked={true}
                  className="answer-input-box"
                  disabled={true}
                  id="1-option-0"
                  name="1-options"
                  onChange={[Function]}
                  type="radio"
                />
                <label
                  className="answer-label"
                  htmlFor="1-option-0"
                >
                  <span
                    className="answer-letter-wrapper"
                  >
                    <button
                      aria-label="Selected Choice a:"
                      className="answer-letter"
                      data-test-id="answer-choice-a"
                      disabled={true}
                      onClick={[Function]}
                    >
                      a
                    </button>
                  </span>
                  <div
                    className="answer-answer"
                  >
                    <span
                      className="answer-content"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "True",
                        }
                      }
                    />
                  </div>
                </label>
              </section>
            </div>
            <div
              className="openstax-answer"
            >
              <section
                className="answers-answer disabled"
                role="region"
              >
                <input
                  checked={false}
                  className="answer-input-box"
                  disabled={true}
                  id="1-option-1"
                  name="1-options"
                  onChange={[Function]}
                  type="radio"
                />
                <label
                  className="answer-label"
                  htmlFor="1-option-1"
                >
                  <span
                    className="answer-letter-wrapper"
                  >
                    <button
                      aria-label="Choice b:"
                      className="answer-letter"
                      data-test-id="answer-choice-b"
                      disabled={true}
                      onClick={[Function]}
                    >
                      b
                    </button>
                  </span>
                  <div
                    className="answer-answer"
                  >
                    <span
                      className="answer-content"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "False",
                        }
                      }
                    />
                  </div>
                </label>
              </section>
            </div>
          </div>
        </div>
        <div
          className="sc-gKXOVf kGsBya step-card-footer"
        >
          <div
            className="points"
          >
            <strong>
              Points: 
              1.0
            </strong>
            <span
              className="attempts-left"
            />
          </div>
          <div
            className="controls"
          >
            <button
              className="sc-eCYdqJ dqiaKp"
              data-test-id="continue-btn"
              onClick={[Function]}
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Exercise with question state data shows a detailed solution 1`] = `
<div
  className="sc-gsnTZi kTXtGE"
  data-task-step-id={1}
>
  <div
    className="sc-bczRLJ hctMxl exercise-step sc-iBkjds Alzlt"
  >
    <div
      className="sc-dkzDqf ifUCJL"
    >
      <div>
        <div
          className="question-info"
        >
          <span>
            Question 1
          </span>
          <span
            className="num-questions"
          >
             / 
            1
          </span>
          <span
            className="separator"
          >
            |
          </span>
          <span
            className="question-id"
          >
            ID: 
            1234@5
          </span>
        </div>
      </div>
      <div>
        <div
          className="points"
        >
          1.0
           Points
        </div>
      </div>
    </div>
    <div
      className="sc-hKMtZM grRsJk"
    >
      <div
        className="step-card-body exercise-context"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Context",
          }
        }
      />
      <div
        className="step-card-body exercise-stimulus"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<b>Stimulus HTML</b>",
          }
        }
      />
      <div
        data-test-id="student-exercise-question"
      >
        <div
          className="sc-jSMfEi kWGKid openstax-question step-card-body"
          data-question-number={1}
          data-test-id="question"
        >
          <div
            className="answers-table"
          >
            <div
              className="openstax-answer"
            >
              <section
                className="answers-answer disabled answer-checked"
                role="region"
              >
                <input
                  checked={true}
                  className="answer-input-box"
                  disabled={true}
                  id="1-option-0"
                  name="1-options"
                  onChange={[Function]}
                  type="radio"
                />
                <label
                  className="answer-label"
                  htmlFor="1-option-0"
                >
                  <span
                    className="answer-letter-wrapper"
                  >
                    <button
                      aria-label="Selected Choice a:"
                      className="answer-letter"
                      data-test-id="answer-choice-a"
                      disabled={true}
                      onClick={[Function]}
                    >
                      a
                    </button>
                  </span>
                  <div
                    className="answer-answer"
                  >
                    <span
                      className="answer-content"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "True",
                        }
                      }
                    />
                  </div>
                </label>
              </section>
            </div>
            <div
              className="openstax-answer"
            >
              <section
                className="answers-answer disabled"
                role="region"
              >
                <input
                  checked={false}
                  className="answer-input-box"
                  disabled={true}
                  id="1-option-1"
                  name="1-options"
                  onChange={[Function]}
                  type="radio"
                />
                <label
                  className="answer-label"
                  htmlFor="1-option-1"
                >
                  <span
                    className="answer-letter-wrapper"
                  >
                    <button
                      aria-label="Choice b:"
                      className="answer-letter"
                      data-test-id="answer-choice-b"
                      disabled={true}
                      onClick={[Function]}
                    >
                      b
                    </button>
                  </span>
                  <div
                    className="answer-answer"
                  >
                    <span
                      className="answer-content"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "False",
                        }
                      }
                    />
                  </div>
                </label>
              </section>
            </div>
          </div>
        </div>
        <div
          className="sc-gKXOVf kGsBya step-card-footer"
        >
          <div
            className="points"
          >
            <strong>
              Points: 
              1.0
            </strong>
            <span
              className="attempts-left"
            />
            <div>
              <strong>
                Detailed solution:
              </strong>
               
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Detailed solution",
                  }
                }
              />
            </div>
          </div>
          <div
            className="controls"
          >
            <button
              className="sc-eCYdqJ dqiaKp"
              data-test-id="continue-btn"
              onClick={[Function]}
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
